File: original-source/Sources/InputMonitor/Pipeline/EventBuffer.swift
import Foundation

/// 事件环形缓冲：定量/定时触发批量 flush，磁盘 I/O 与事件频率解耦
final class EventBuffer {
    private let capacity: Int
    private let flushInterval: TimeInterval
    private var buffer: [InputEventRecord] = []
    private var timer: DispatchSourceTimer?
    private let queue = DispatchQueue(label: "com.local.inputmonitor.buffer")

    /// flush 回调（在 queue 上触发）
    var onFlush: (([InputEventRecord]) -> Void)?

    init(capacity: Int = 500, flushInterval: TimeInterval = 5) {
        self.capacity = capacity
        self.flushInterval = flushInterval
    }

    func start() {
        let timer = DispatchSource.makeTimerSource(queue: queue)
        timer.schedule(deadline: .now() + flushInterval, repeating: flushInterval)
        timer.setEventHandler { [weak self] in
            self?.flushLocked()
        }
        timer.resume()
        self.timer = timer
    }

    func stop() {
        timer?.cancel()
        timer = nil
        queue.sync { flushLocked() }
    }

    func append(_ record: InputEventRecord) {
        queue.async {
            self.buffer.append(record)
            if self.buffer.count >= self.capacity {
                self.flushLocked()
            }
        }
    }

    func flush() {
        queue.async { self.flushLocked() }
    }

    private func flushLocked() {
        guard !buffer.isEmpty else { return }
        let batch = buffer
        buffer.removeAll(keepingCapacity: true)
        onFlush?(batch)
    }
}

File: current-integration/src/InputMonitor.Core/EventBuffer.cs
namespace InputMonitor.Core;

public sealed class EventBuffer : IDisposable
{
    private readonly int _capacity;
    private readonly TimeSpan _flushInterval;
    private readonly object _gate = new();
    private readonly List<InputEventRecord> _buffer = [];
    private readonly Timer _timer;
    private bool _disposed;

    public Action<IReadOnlyList<InputEventRecord>>? OnFlush { get; set; }

    public EventBuffer(int capacity = 500, TimeSpan? flushInterval = null)
    {
        _capacity = capacity;
        _flushInterval = flushInterval ?? TimeSpan.FromSeconds(5);
        _timer = new Timer(_ => Flush(), null, _flushInterval, _flushInterval);
    }

    public void Append(InputEventRecord record)
    {
        IReadOnlyList<InputEventRecord>? batch = null;
        lock (_gate)
        {
            if (_disposed)
            {
                return;
            }

            _buffer.Add(record);
            if (_buffer.Count >= _capacity)
            {
                batch = DrainLocked();
            }
        }

        if (batch is not null)
        {
            OnFlush?.Invoke(batch);
        }
    }

    public void Flush()
    {
        IReadOnlyList<InputEventRecord> batch;
        lock (_gate)
        {
            batch = DrainLocked();
        }

        if (batch.Count > 0)
        {
            OnFlush?.Invoke(batch);
        }
    }

    public void Dispose()
    {
        lock (_gate)
        {
            if (_disposed)
            {
                return;
            }

            _disposed = true;
        }

        _timer.Dispose();
        Flush();
    }

    private IReadOnlyList<InputEventRecord> DrainLocked()
    {
        if (_buffer.Count == 0)
        {
            return [];
        }

        var batch = _buffer.ToArray();
        _buffer.Clear();
        return batch;
    }
}
